helix-event/src/cancel.rs

```diff
@@ -1,19 +1,285 @@
+            assert_ne!(num_running, 0);
+            let off = 1 << 32;
+            let res = self
+                .state
+                .compare_exchange_weak(state, state - off, Relaxed, Relaxed);
+            match res {
+                Ok(_) => break,
+                Err(new_state) => state = new_state,
+            }
+        }
+    }
+}
+
+// this intentionally doesn't implement clone and requires amutable reference
+// for cancelation to avoid races (in inc_generation)
+
+/// A task controller allows managing a single subtask enabling the contorller
+/// to cancel the subtask and to check wether it is still running. For efficency
+/// reasons the controller can be reused/restarted, in that case the previous
+/// task is automatically cancelled.
+///
+/// If the controller is dropped the subtasks are automatically canceled.
+#[derive(Default, Debug)]
+pub struct TaskController {
+    shared: Arc<Shared>,
+}
+
+impl TaskController {
+    pub fn new() -> Self {
+        TaskController::default()
+    }
+    /// Cancels the active task (handle)
+    ///
+    /// returns wether any tasks were still running before the canellation
+    pub fn cancel(&mut self) -> bool {
+        self.shared.inc_generation(0).1 != 0
+    }
+
+    /// checks wether there are any task handles
+    /// that haven't been dropped (or canceled) yet
+    pub fn is_running(&self) -> bool {
+        self.shared.num_running() != 0
+    }
+
+    /// Starts a new task and cancels the previous task (handles)
+    pub fn restart(&mut self) -> TaskHandle {
+        TaskHandle {
+            generation: self.shared.inc_generation(1).0,
+            shared: self.shared.clone(),
+        }
+    }
+}
+
+impl Drop for TaskController {
+    fn drop(&mut self) {
+        self.cancel();
+    }
+}
+
+/// A handle that is used to link a task with a task controller, it can be
+/// used to cancel async futures very efficently but can also be checked for
+/// cancaellation very quickly (single atomic read) in blocking code. The
+/// handle can be cheaply cloned (referenced counted).
+///
+/// The TaskController can check wether a task is "running" by inspecting the
+/// refcount of the (current) tasks handeles. Therefore, if that information
+/// is important ensure that the handle is not dropped until the task fully
+/// completes
+pub struct TaskHandle {
+    shared: Arc<Shared>,
+    generation: u32,
+}
+
+impl Clone for TaskHandle {
+    fn clone(&self) -> Self {
+        self.shared.inc_running(self.generation);
+        TaskHandle {
+            shared: self.shared.clone(),
+            generation: self.generation,
+        }
+    }
+}
+
+impl Drop for TaskHandle {
+    fn drop(&mut self) {
+        self.shared.dec_running(self.generation);
+    }
+}
+
+impl TaskHandle {
+    /// waits until [`TaskController::cancel`] is called for the corresponding
+    /// [`TaskController`]. Immidietly returns if `cancel` was already called since
+    pub async fn canceled(&self) {
```

Comment:
```suggestion
    pub async fn wait_until_canceled(&self) {
```

Probably more descriptive?

Reply:
I followwd tokios naming convention here (for example a Notify as a notified() future and not a wait_notified future). If you call this `token.canceled().await` the wait part is clear (.await). I don't feel too strongly, if you think it's much better than feel free to change it

Reply:
Ah haven't seen it that way, ok that makes sense then. No I don't have strong feelings for a rename either...